src/Chapter3.hs

```diff
@@ -374,6 +386,31 @@ after the fight. The battle has the following possible outcomes:
 
 -}
 
+data Creature = Creature
+  {
+    health :: Int,
+    attack :: Int,
+    gold :: Int
+  }
+
+
+-- turn 1 is Knight, 0 is Monster
+fight :: Creature -> Creature -> Int -> Int
+fight k m turn
+  | (health k) <= 0 && (health m) > 0 = -1
+  | (health k) > 0 && (health m) <= 0 = (gold k) + (gold m)
+  | (health k) <= 0 && (health m) <= 0 = (gold k)
+  | turn == 1 = fight k m2 0
+  | turn == 0 = fight k2 m 1
```

Comment:
As we mentioned, the compiler in Haskell is very attentive to the exhaustive pattern-matching. And here it would warn you that `Pattern matching is not exhaustive`, as the guards have quite complicated logic, and the compiler won't be able to prove that it covers all the cases. At least, you don't cover the case when `turn == 3` or any other number. 
I understand that `turn` could never be 3, but as per type, it could. So it would be better to use more representable type for these 2 states. 
This pattern is called "Make illegal states unrepresentable", and you can read more about it here:

* https://kowainik.github.io/posts/haskell-mini-patterns#make-illegal-states-unrepresentable

Because of that, you will need to add another guard – `| otherwise = ...`, to tell the compiler, that your pattern matching is exhaustive 🙂